src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/SwerveStateHelper.java

```diff
@@ -0,0 +1,115 @@
+    private final Supplier<Optional<Translation2d>> noteTranslationSupplier;
+
+    public SwerveStateHelper(Supplier<Pose2d> robotPoseSupplier, Supplier<Optional<Translation2d>> noteTranslationSupplier,
+            Swerve swerve) {
+        this.robotPoseSupplier = robotPoseSupplier;
+        this.swerve = swerve;
+        this.swerveConstants = swerve.getConstants();
+        this.noteTranslationSupplier = noteTranslationSupplier;
+    }
+
+    public Translation2d getRotationAxis(RotateAxis rotationAxisState) {
+        return switch (rotationAxisState) {
+            case FRONT_LEFT_MODULE -> swerveConstants.LOCATIONS[ModuleUtils.ModuleName.FRONT_LEFT.getIndex()];
+            case FRONT_RIGHT_MODULE -> swerveConstants.LOCATIONS[ModuleUtils.ModuleName.FRONT_RIGHT.getIndex()];
+            case BACK_LEFT_MODULE -> swerveConstants.LOCATIONS[ModuleUtils.ModuleName.BACK_LEFT.getIndex()];
+            case BACK_RIGHT_MODULE -> swerveConstants.LOCATIONS[ModuleUtils.ModuleName.BACK_RIGHT.getIndex()];
+            case MIDDLE_OF_ROBOT -> new Translation2d();
+        };
+    }
+
+    public ChassisSpeeds applyAimAssistOnInputsSpeeds(AimAssist aimAssistState, ChassisSpeeds inputSpeeds) {
+        return switch (aimAssistState) {
+            case SPEAKER -> handleSpeakerState(
+                    inputSpeeds,
+                    robotPoseSupplier,
+                    robotPose -> new Translation3d(0,0,0).toTranslation2d().minus(robotPose.getTranslation()).getAngle(),
+                    swerveConstants
+            );
+            case AMP -> handleAmpState(
+                    inputSpeeds,
+                    robotPoseSupplier,
+                    (robotPose -> Rotation2d.fromDegrees(90)),
+                    swerveConstants
+            );
+            case NOTE -> handleNoteAssistState(
+                    inputSpeeds,
+                    robotPoseSupplier,
+                    noteTranslationSupplier
+            );
+            case NONE -> inputSpeeds;
+            default -> inputSpeeds;
+        };
+    }
+
+
+    private ChassisSpeeds handleNoteAssistState(ChassisSpeeds inputSpeeds, Supplier<Pose2d> robotPoseSupplier,
+            Supplier<Optional<Translation2d>> noteTranslationSupplier) {
+        if (noteTranslationSupplier.get().isEmpty()) {
+            return inputSpeeds;
+        }
+        Rotation2d robotRotation = robotPoseSupplier.get().getRotation().unaryMinus();
+
+        Translation2d noteTranslation = noteTranslationSupplier.get().get();
+        Translation2d normalizedNoteTranslation = noteTranslation.minus(robotPoseSupplier.get().getTranslation());
+        Translation2d rotatedNoteTranslation = normalizedNoteTranslation.rotateBy(robotRotation);
+
+        Logger.recordOutput("note pose", new Pose2d(noteTranslation, new Rotation2d()));
+
+        double wantedHorizontalVelocity = swerveConstants.yMetersPIDController().calculate(0, rotatedNoteTranslation.getY())
+                + inputSpeeds.vyMetersPerSecond;
```

Comment:
also this line same as above

Reply:
like make a function that gets 2 doubles and adds them? 🤣

Reply:
i think people can understand basic math